h2o-core/src/main/java/hex/AUC2.java

```diff
@@ -279,31 +291,53 @@ public void perRow(double pred, int act, double w ) {
           _ths[idx] = _ths[idx] + (pred-_ths[idx])/oldk;
           _sqe[idx] = _sqe[idx] + d0;
           assert ssx == find_smallest();
-          return;
         }
+      } else {  // insert this row
+
+        // Must insert this point as it's own threshold (which is not insertion
+        // point), either because we have too few bins or because we cannot
+        // instantly merge the new point into an existing bin.
+        if (idx == _ssx) _ssx = -1;  // Smallest error becomes one of the splits
+        else if (idx < _ssx) _ssx++; // Smallest error will slide right 1
+
+        // Slide over to do the insert.  Horrible slowness.
+        System.arraycopy(_ths, idx, _ths, idx + 1, _n - idx);
+        System.arraycopy(_sqe, idx, _sqe, idx + 1, _n - idx);
+        System.arraycopy(_tps, idx, _tps, idx + 1, _n - idx);
+        System.arraycopy(_fps, idx, _fps, idx + 1, _n - idx);
+        // Insert into the histogram
+        _ths[idx] = pred;         // New histogram center
+        _sqe[idx] = 0;            // Only 1 point, so no squared error
+        if (act == 0) {
+          _tps[idx] = 0;
+          _fps[idx] = w;
+        } else {
+          _tps[idx] = w;
+          _fps[idx] = 0;
+        }
+        _n++;
       }
 
-      // Must insert this point as it's own threshold (which is not insertion
-      // point), either because we have too few bins or because we cannot
-      // instantly merge the new point into an existing bin.
-      if( idx == _ssx ) _ssx = -1;  // Smallest error becomes one of the splits
-      else if( idx < _ssx ) _ssx++; // Smallest error will slide right 1
-
-      // Slide over to do the insert.  Horrible slowness.
-      System.arraycopy(_ths,idx,_ths,idx+1,_n-idx);
-      System.arraycopy(_sqe,idx,_sqe,idx+1,_n-idx);
-      System.arraycopy(_tps,idx,_tps,idx+1,_n-idx);
-      System.arraycopy(_fps,idx,_fps,idx+1,_n-idx);
-      // Insert into the histogram
-      _ths[idx] = pred;         // New histogram center
-      _sqe[idx] = 0;            // Only 1 point, so no squared error
-      if( act==0 ) { _tps[idx]=0; _fps[idx]=w; }
-      else         { _tps[idx]=w; _fps[idx]=0; }
-      _n++;
-      if( _n > _nBins )         // Merge as needed back down to nBins
-        mergeOneBin();          // Merge best pair of bins
+      // Merge duplicate rows in _ths.  May require many merges.  May or may  not cause reproducibility issue
+      removeDupsShrink(_nBins, false);
     }
 
+    // Merge duplicate rows in all 4 arrays.
+    public void removeDupsShrink(int maxBinSize, boolean setrError) {
```

Comment:
This method is called from perRow(..). It is important to note that
- there can be at most one duplicate,
- only one non-duplicate bin will actually be merged.

The while loops are therefore not necessary.